pkg/collector/corechecks/snmp/internal/report/report_bandwidth_usage.go

```diff
@@ -97,18 +98,76 @@ func (ms *MetricSender) sendBandwidthUsageMetric(symbol profiledefinition.Symbol
+	err = ms.calculateRate(interfaceID, ifSpeed, usageValue, usageName, tags)
+	if err != nil {
+		return err
 	}
+	return nil
+}
+
+/*
+calculateRate is responsible for checking the state for previously seen metric sample to generate the rate from.
+If the ifSpeed has changed for the interface, the rate will not be submitted (drop the previous sample)
+*/
+func (ms *MetricSender) calculateRate(interfaceID string, ifSpeed uint64, usageValue float64, usageName string, tags []string) error {
+	ms.interfaceRateMap.mu.RLock()
+	interfaceRate, ok := ms.interfaceRateMap.rates[interfaceID]
+	ms.interfaceRateMap.mu.RUnlock()
+
+	// current data point has the same interface speed as last data point
+	if ok && interfaceRate.ifSpeed == ifSpeed {
+		// calculate the delta, taken from pkg/metrics/rate.go
+		// https://github.com/DataDog/datadog-agent/blob/ecedf4648f41193988b4727fc6f893a0dfc3991e/pkg/metrics/rate.go#L37
+		currentTimestamp := TimeNow()
+		delta := (usageValue - interfaceRate.previousSample) / (currentTimestamp - interfaceRate.previousTs)
+		// update the map previous as the current for next rate
+		interfaceRate.previousSample = usageValue
+		interfaceRate.previousTs = currentTimestamp
+		ms.interfaceRateMap.mu.Lock()
+		ms.interfaceRateMap.rates[interfaceID] = interfaceRate
+		ms.interfaceRateMap.mu.Unlock()
+
+		if delta < 0 {
+			return fmt.Errorf("Rate value for device/interface %s is negative, discarding it", interfaceID)
+		}
 
-	ms.sendMetric(sample)
+		// create the gauge metric sample
+		sample := MetricSample{
+			value:      valuestore.ResultValue{SubmissionType: profiledefinition.ProfileMetricTypeGauge, Value: delta},
+			tags:       tags,
+			symbol:     profiledefinition.SymbolConfig{Name: usageName + ".rate"},
+			forcedType: profiledefinition.ProfileMetricTypeGauge,
+			options:    profiledefinition.MetricsConfigOption{},
+		}
+		// send the metric
+		ms.sendMetric(sample)
+	} else {
+		// otherwise, no previous data point / different ifSpeed - make new entry for interface
+		ms.interfaceRateMap.mu.Lock()
+		ms.interfaceRateMap.rates[interfaceID] = InterfaceRate{
+			ifSpeed:        ifSpeed,
+			previousSample: usageValue,
+			previousTs:     TimeNow(),
+		}
+		ms.interfaceRateMap.mu.Unlock()
+		// do not send a sample to metrics, send error for ifSpeed change (previous entry conflicted)
+		if ok {
+			log.Infof("ifSpeed changed from %d to %d for device and interface %s, no rate emitted", interfaceRate.ifSpeed, ifSpeed, interfaceID)
+		}
+	}
 	return nil
 }
 
+// TimeNow is the unix time to use for rate (delta) calculations
+var TimeNow = timeNowNano
+
+// Helper function to determine the timestamp for the rate metric sample, lifted from pkg/aggregator
+// https://github.com/DataDog/datadog-agent/blob/ecedf4648f41193988b4727fc6f893a0dfc3991e/pkg/aggregator/aggregator.go#L96
+func timeNowNano() float64 {
+	return float64(time.Now().UnixNano()) / float64(time.Second) // Unix time with nanosecond precision
+}
```

Comment:
Shall we mock the "time now" like this https://github.com/DataDog/datadog-agent/blob/ffc04209c57eea33fd845db150766b5896e423e8/pkg/collector/corechecks/snmp/internal/devicecheck/devicecheck.go#L51-L52

instead of mocking the whole:

```
func timeNowNano() float64 {
	return float64(time.Now().UnixNano()) / float64(time.Second) // Unix time with nanosecond precision
}
```

I think the mock as little as possible, that way the test will have better coverage. For example here, the content of `timeNowNano()` doesn't seems to be tested.

Reply:
i see, so something like:

```
var timeNow = time.Now.UnixNano()
```

at this point i won't need to have the function if i convert to float 64 later with the conversion to seconds when i need to utilize it